src/Base/ObjectDependencyManager.cs

```diff
@@ -141,36 +141,44 @@ public static void CleanUp(object objToRemove)
             // this list will hold all keys they has to be removed
             List<object> keysToRemove = new List<object>();
 
+            // This list will hold all references which should be removed.
+            // It's one for all keys for reduce memory allocations.
+            List<WeakReference> deadReferences = new List<WeakReference>();
+
             // step through all object dependenies
-            foreach (KeyValuePair<object, List<WeakReference>> kvp in internalList)
+            foreach (KeyValuePair<object, HashSet<WeakReference>> kvp in internalList)
             {
-                // step recursive through all weak references
-                for (int i = kvp.Value.Count - 1; i >= 0; i--)
+                foreach (var target in kvp.Value)
                 {
-                    // if this weak reference is no more alive, remove it
-                    var targetReference = kvp.Value[i].Target;
+                    var targetReference = target.Target;
                     if (targetReference == null)
-                    {
-                        kvp.Value.RemoveAt(i);
-                    }
+                        deadReferences.Add(target);
                 }
 
-                // if the list of weak references is empty, temove the whole entry
-                if (kvp.Value.Count == 0)
+                if (deadReferences.Count > 0)
                 {
-                    keysToRemove.Add(kvp.Key);
+                    // if the all of weak references is empty, remove the whole entry
+                    if (deadReferences.Count == kvp.Value.Count)
+                    {
+                        keysToRemove.Add(kvp.Key);
+                    }
+                    else
+                    {
+                        foreach (var deadReference in deadReferences)
+                        {
+                            kvp.Value.Remove(deadReference);
```

Comment:
does it make sense to add in Line 160 a var value = kvp.Value.
Because in this look you have always this lookup and you can reuse it also in 161 for the kvp.Value.Count